### DISCUSSION TITLE
Issues with LPS and Drones Crashing

### DISCUSSION THREAD

**Discussion by Ometz** [bitcraze/discussions#746] [Q&A] [answered]
Hi,
 
I am having issues causing my drones to crash. I think it is to do with the LPS system.
I have 2 sets of data here to show you.

Results set 1 is here:

Setup: Ubuntu 20.04 running natively on a desktop. **Crazyswarm2** (ros2)
Results set 1 obtained testing control issues with 2 different CrazyFlies.
Radio frequency was 100 with mbps connection
Mistake: Forgot to switch to PID controller however previous tests with PID Controller give same results
Crazyflies:
CF02 HASPROPS
CF06 HAS NO PROPS (can't fly)
 
Two tests conducted with helloworld.py hover,
Two tests conducted with static propellerless drone

Test 1 (Figure 1):
CF02 with Flowdeck only and lps deck LPS TURNED OFF.
Results: Successful hover and landing
![Test1Result1](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/16857884-e310-418b-8130-a32f1ef86301)
Figure 1: position and orientation plots for test 1

Test 2 (Figure 2):
CF02 with flowedeck and lps deck LPS TURNED ON.
Results: Failure, take off, begin drifting and crash
![Test1Result2](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/61dc9884-8de8-40f5-94b1-df021257da04)
Figure 2: position and orientation plots for test 2

 
At this point the battery to board connection on CF02 broke forcing me to use CF06
Test 3 (Figure 3):
CF06 with Flowdeck only.
Results: data recorded
![Test1Result3](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/59389263-d3da-479c-87ee-ef8ca2016b29)
Figure 3: position and orientation plots for test 3

 
Test 4 (Figure 4):
CF06 with Lps.
Results: data recorded
![Test1Result4](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/a5e2dcab-0dda-4a25-853f-6900a146a73b)
Figure 4: position and orientation plots for test 4

**Results 2 setup:**
Razer Blade 15 advanced (2020)
Bitcraze VM,
crazyflie_lib_python with Ros1 wrapper just to get bag data
Crazyflie with only LPS deck **(now flowdeck attached)** flying a test in each LPS mode
 
Results part 2 shows logging of the Kalman estimates in comparison to the State estimates. My intent was just to check if these were actually equivalent and they are not. I was also checking each mode of the LPS on the Bitcraze VM itself, **all resulting in a crash** although the TWR took off slower and more tentatively. I think this might have been due to it thinking it was already off the ground however, and since I am doing swarming I wouldn't really use this mode anyhow. PID was being used as the controller this time I believe.

Result 1 (figure 5): TDOA3
![Test2Result1_TDOA3](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/8dec1370-97f2-490a-b0b3-e06207db6993)
Figure 5: X position of Drone using TDOA3

Result 2 (figure 6): TDOA2
![Test2Result2_TDOA2](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/56039eeb-ba07-42aa-8a47-94c74d491588)
Figure 6: X position of Drone using TDOA2

Result 3 (Figure 7): TWR
![Test2Result3_TWR](https://github.com/bitcraze/crazyflie-lib-python/assets/12114988/c165fd90-e829-4d10-8668-c04519397abe)
Figure 7: X position of Drone using TWR

I would like to figure out what is wrong with the nodes here as we have many of those purchased as well as many nodes. 
One additional solution to these issues might be to just set the global position every so often some sort of ground truth using the lps and have the flowdeck do the rest. I only have 2 flowdecks however and I am hoping to work with swarms. That being said, I was still wondering how I could try that.

Any help is appreciated.
Thanks

**Comment by krichardsson** [comment, marked as answer]
The first thing to check is that you place the[ Crazyflie in the right orientation ](https://www.bitcraze.io/documentation/tutorials/getting-started-with-flying-using-lps/#restart-the-crazyflie) before taking off using the Loco system. The loco system can not measure yaw and it is important that the orientation is correct.

Have you set up the system using the python client, as described in the [getting started tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/)? Does the position estimate looks reasonable in the client? Are all anchors receied?

> Results part 2 shows logging of the Kalman estimates in comparison to the State estimates. My intent was just to check if these were actually equivalent and they are not. 

I think the difference you see is because they are sampled at different times. There is no guarantee for exactly when the logging framework is sampling data and it depends on a number of factors.

**Reply by Ometz** [reply]
Hi,
 
Thanks for your response, and sorry for my late one. I ran some more tests, making sure to position the crazyflie correctly as you suggested and I definitely improved my results by being more consistent with the alignment. In the Bitcraze VM with just cflib python and my ros wrapper there are no longer any crashes and the data looks relatively good although with just the lps deck and no flowdeck it has a hard time determining its height. Crazyswarm2 seems to have a harder time, though I do have a run with no crashes to show you and it has definitely improved overall.

Right now my issues with the lps are that the nearer the drones get to the edges of my setup the worse the estimates get, especially around the origin. Because of the small room I've been given this basically leaves me with a 1 meter diameter cylinder of good estimates. It also has trouble actually getting zero height correct. I'm not sure if there is a way I can actually tell it when it is there. Right now I have my node zero set at xyz = (0,0,0.3) in meters. There is slightly under a foot of space between that and the actual corner of the room. I could make it more like (0.25,0.25,0.3), might that help at all with my estimates? Any other suggestions for lps accuracy would also be appreciated.
 
In any case here are my results implementing your suggestions. As stated there are no longer crashes with the crazyflie-lib python service running take off and land. I should note that the expected behavior is to take off to 0.5 meters, hover at that height, then move up 0.2 meters, then move down 0.2 meters.
 
 
 
**Results part 1 (cflib ros wrapper):**
 
![figure 1](https://github.com/bitcraze/discussions/assets/12114988/5e33ed5f-a4cf-44e4-809f-245b12ddd28c)
**Figure 1: Test with CF12 no flowdeck attached, only lps deck**

![side_by_side1](https://github.com/bitcraze/discussions/assets/12114988/b5d738b0-d93f-4640-99da-078ad96b1380)
**Figure 2 and 2B: Test with CF8 no flowdeck attached, only LPS deck, uncropped and cropped**
 
![figure3](https://github.com/bitcraze/discussions/assets/12114988/ed34af92-8e1b-4259-9a5a-63098b8d7a2b)
**Figure 3: Test with CF06 Flowdeck and Lps decks attached (better estimates)**
 
 
**Results part 2 (crazyswarm2):**
 
![side_by_side_swarm2](https://github.com/bitcraze/discussions/assets/12114988/97fb8a1c-84ae-4c76-8da6-b911b50a64f0)
**Figure 4 and 4b: Side by side results for Cf6 (lps and flowdeck) and Cf8 (only lps) CrazySwarm 2. Caution that y axes aren't the same here**
 
Both drones here reached there commanded positions of xyz (1.8,1.8,1) for cf06 and xyz (1.4,1.4,0.4) for cf08. It is interesting that even though the estimates give a smoother curve with the flowdeck, it seems to be oscillating around more rather than converging at its commanded position. Not sure why this is.
 
In any case, thank you for your help! It is very much appreciated.

**Comment by krichardsson** [comment]
The Z-estimate is often a problem in Loco system and unfortunately we do not fully understand why. The main underlying problem is that the radiation pattern of the UWB antenna is not uniform and distance measurements to an anchor changes a lot (around 50 cm) depending on the orientation. What I don't understand is that it seems to affect the Z-estimate differently compared to X and Y.

You can play around with the standard deviation of the UWB measurements that are sent to the kalman estimator. You set the std through the [tdoa2.stddev](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#tdoa2stddev) or [tdoa3.stddev](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#tdoa3stddev) parameter. I think it is 0.15 by default, you can try to increase it to maybe 0.3. If you also have a flow deck you can probably bump it up even more.

**Reply by Ometz** [reply]
Thanks for the response, I will experiment with it as you have suggested.
